File: Algorithms/src/linked_list.cpp
#include <iostream>
using namespace std;

struct node{
    int data;
    node* next;
};

node* initialize_node(int data);
bool isEmpty(node* head);
int count_sum(node* head);
int get_element_length(node* head);
void insert_at_head(node* &head, int data);
void insert_at_queue(node* &head, int data);
bool is_in_list(node* head,int target_data);
node* find_element(node* head,int target_data);
void print_list(node* head);
void delete_element(node* &head,int val);
node* invert_list(node* head);

int main(){

    // Run the wanted script here
    return 0;
            
}

node* initialize_node(int data){
    node* tmp;
    tmp->data = data;
    tmp->next = NULL;
    return;
}

bool is_empty(node* head){
    return (head == NULL);
}

int get_element_length(node* head){
    if(head->next == NULL)
        return 1;
    return get_element_length(head->next)+1;
}

int count_sum(node* head){
    if(head->next == NULL)
        return head->data;
    return count_sum(head->next)+head->data;
}

void insert_at_head(node* &head, int data){

    node* tmp;
    tmp = initialize_node(data);
    tmp->next = head;
    head = tmp;

}

void insert_at_queue(node* &head, int data){
    
    node* new_node = initialize_node(data);
    while(head->next != NULL)
        head = head->next;
    head->next = new_node;

}

bool is_in_list(node* head,int target_data){

    while(head != NULL){
        if(head->data == target_data)
            return true;
        head = head->next;
    }
    return false;

}

node* find_element(node* head,int val){

    while(head->next != NULL){
        if(head->data == val)
            return head;
        head = head->next;
    }
    return NULL;

}

void print_list(node* head){
    while(head->next != NULL){
        cout<<head->data<<" ";
    }
}

void delete_element(node* &head, int val){

    //Case where the first element is the element to be deleted
    if(head->data == NULL){
       head = head->next; 
        return;
    }

    node* cur = head;
    node* prev;
    while(cur->next != NULL){ //Case where the element is in the middle of the list
        if(cur->data == val){
            prev->next = cur->next;
            cur = cur->next;
            return;
        }
        prev = cur;
        cur = cur->next;
    }

    //Case where the element is the last element of the function
    if(cur->next == NULL && cur->data == val){
        prev->next = NULL;
        cur = prev;
    }
    
}

node* invert_list(node* head){

    node* prev = NULL, *cur = NULL;

    while (head != NULL){
        prev = cur;
        head = head->next;
        cur = head;
        cur->next = prev;
    }
    head = cur;
    return head;

}

node* delete_list(node* head){

    node* tmp;
    while(head != NULL){
        tmp = head;
        head = head->next;
        delete tmp;
    }

}

File: Algorithms/python/linked_list.py
from asyncio.windows_events import NULL
from curses import noecho


class Node:

    #Initializing the new Node
    def __init__(self, data, next=None): 
        self.data = data
        self.next = next

class LinkedList:

    # Initializing the new Linked List
    def __init__(self):  
        self.head = None

    # Check if Linked list is empty
    def is_empty(self):
        return self.head and self.data

    # Return Number of elements in the linked list
    def get_length(self):
        if self.is_empty(): return 0
        ctr = 0
        while(self.head.next):
            ctr += 1
            self.head = self.head.next
        return ctr

    # Return sum of all nodes elements
    def count_value(self):
        if self.is_empty(): return 0
        sum = 0
        while(self.head.next):
            sum += self.head.data
            self.head = self.head.next
        return sum
    
    # Insert node to the queue of the linked list
    def insert_at_head(self,data):
        newNode = Node(data)
        if self.head:
            newNode.next = self.head
        self.head = newNode
        return self

    # Insert node to the queue of the linked list
    def insert_at_queue(self,data):
        newNode = Node(data)
        if(self.head):
            current = self.head
            while(current.next):
                current = current.next
            current.next = newNode
        else:
            self.head = newNode

    # Check if the element is in the list
    def is_in_list(self,data):
        if self.head.is_empty(): 
            return

        while self.head.next :
            if self.head.data == data:
                return True
            self.head = self.head.next
        return False

    # Find an element and return a node
    def find_element(self,data):
        if not self.head.is_empty():
            return
        
        while self.head.next != None:
            if self.head.data == data:
                return self.head
        return NULL


    def delete_element(self,data):

        if self.head.is_empty():
            return self.head
        
        if self.head.data:
            tmp = self.head.next
            del self.head
            self.head = tmp
            return self.head

        prev = self.head
        cur = self.head.next
        while cur.next != None:
            if cur.data == data:
                prev.next = cur.next
                del cur
                cur = prev.next
                return self.head
            prev = cur
            cur = cur.next

        if cur.data == data:
            prev.next = None
            del cur

        return self.head
        

    # Display Linked List
    def display(self):
        if not self.head:
            print("Empty Linked list!")
            return
        while self.head.next:
            print(self.head.data)
            self = self.head.next


def main():
    l = LinkedList()
    

if __name__ == '__main__':
    main()